### DISCUSSION TITLE
ParamArrayAttribute not generated for local function with params argument

### DISCUSSION THREAD

**Discussion by Joy-less** [dotnet/csharplang#7748] [Q&A]
I'm trying to check whether an argument for a function is `params` using reflection. It seems like the only way to do this is find a `ParamArrayAttribute`. However, it seems this is not generated for local functions! Is there another way?

```cs
using System;
using System.Reflection;
using System.Linq;
					
public class Program {
	static void ClassMethod(params object[] Args) {

	}
	public static void Main() {
		static void LocalMethod(params object[] Args) {
			
		}
		
		Console.WriteLine(((Delegate)ClassMethod).Method.GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
		Console.WriteLine(((Delegate)ClassMethod).Method.GetParameters()[0].GetCustomAttribute<ParamArrayAttribute>() is not null);
		Console.WriteLine(((Delegate)ClassMethod).Method.GetParameters()[0].CustomAttributes.Count());

		Console.WriteLine(((Delegate)LocalMethod).Method.GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
		Console.WriteLine(((Delegate)LocalMethod).Method.GetParameters()[0].GetCustomAttribute<ParamArrayAttribute>() is not null);
		Console.WriteLine(((Delegate)LocalMethod).Method.GetParameters()[0].CustomAttributes.Count());
	}
}
```

Output (tested only for .NET 7 and .NET 8):
```
True
True
1
False
False
0
```

**Comment by HaloFour** [comment]
Additional metadata isn't necessary in the case of local functions as the member isn't accessible outside of the assembly and the compiler is well aware that it accepts params.

Why would you need to know this through reflection?  How the method is encoded should be treated as implementation details and anything not explicitly stated in the specification should never be relied upon.

**Reply by Joy-less** [reply]
The reason is that neither `Delegate.Method.Invoke` nor `Delegate.DynamicInvoke` support `params`. So params arguments need to be condensed into an array. Take the following example:

```cs
using System;
using System.Reflection;
using System.Linq;
					
public class Program {
	public static void Main() {
		static void MyInteropMethod(params object[] Args) {
			
		}
		
		DynamicMethod MyDynamicMethod = new(MyInteropMethod);
		
		object[] MyDynamicArguments = new object[] {3, 5};
		
		MyDynamicMethod.Invoke(MyDynamicArguments);
	}
	class DynamicMethod {
		public readonly Delegate MethodDelegate;
		
		public DynamicMethod(Delegate method_delegate) {
			MethodDelegate = method_delegate;
		}
		public void Invoke(object[] Arguments) {
			MethodDelegate.Method.Invoke(MethodDelegate.Target, Arguments);
		}
	}
}
```
Output:
```
System.Reflection.TargetParameterCountException: Parameter count mismatch.
```

**Reply by HaloFour** [reply]
Sure, but `params` doesn't affect the signature of the method, it always accepts an array and that's how it would appear via reflection.

**Reply by Joy-less** [reply]
But it doesn't accept a non-array, which defeats the purpose of `params`.

**Reply by HaloFour** [reply]
> But it doesn't accept a non-array, which defeats the purpose of `params`.

The purpose of `params` is to be a hint to a supporting compiler to allow it to flatten the arguments.  It's never affected how the runtime interprets the method.

**Reply by Joy-less** [reply]
So why is `ParamArrayAttribute` present for class methods?

**Reply by HaloFour** [reply]
As a hint to supporting compilers that it can allow `params` syntax there.  It's purely syntax candy, the compiler emits those arguments as an array as that is what the method accepts.

**Reply by Joy-less** [reply]
Well I'm making a supporting compiler, and it won't let me check whether it can allow `params` syntax for a local method.

**Reply by HaloFour** [reply]
Local methods aren't publicly exposed, you should never be calling them directly from your compiler.

**Reply by Joy-less** [reply]
They are publicly exposed if they are passed as an argument to a function, see my first reply. It seems unnecessary to block local methods being passed just because `params` is not supported.

**Reply by HaloFour** [reply]
No, the delegate is exposed.  The underlying method target isn't public, and when invoked via delegate or via reflection you need to pass the array explicitly anyway.  The `ParamArrayAttribute` is explicitly for a compiler calling a method directly.

You should never be depending on the signature of a local function anyway, it will change depending on what the local function does with the enclosing scope and what else is done in that enclosing scope.  What the C# compiler emits is entirely an implementation detail.

**Reply by Joy-less** [reply]
You've just made that assertion, without explaining why it should work that way. Supporting either the finding of `ParamArrayAttribute` or `Invoke` that follows `params` is required to fully interop a local method with an interpreter/compiler.

**Reply by HaloFour** [reply]
Because the local method should not be invoked by any compiler except the C# compiler, and only within the method in which it is declared.  Outside compilers shouldn't need to know nor care that the method exists.  If a reference to the method is captured as a delegate and passed around to other methods those other methods should interact with it based on the explicit signature of the delegate, not the underlying target.

**Reply by Joy-less** [reply]
So what's your alternative? Moving local functions outside into the class definition for no good reason other than to support `params` (and lose variable capturing in the process, requiring you to create a whole wrapper class for it)?

**Reply by HaloFour** [reply]
You'll note that your example fails with both local functions and normal methods,  because `ParamArrayAttribute` doesn't have any effect on `Delegate.Invoke`.  You're still required to pass the `params` as an array via reflection.

**Reply by Joy-less** [reply]
I know they both fail via `Delegate.Invoke`, but at least with `ParamArrayAttribute` I can verify that it is params and convert it to an array automatically.

**Reply by HaloFour** [reply]
I'd suggest that relying on anything except the signature of the delegate itself, such as trying to negotiate the target via reflection, is a brittle strategy.

**Reply by Joy-less** [reply]
I'm making a Ruby interpreter, which is a language with splat arguments (`*`) that function exactly like `params` in C#. To interop C# methods without pain, checking for `params` is not a "brittle strategy" but by definition the only way to implicitly interop. Why should the user have to specify the last argument is `params` (with a bool) when it should be done automatically by the compiler (and normally is)?

**Reply by HaloFour** [reply]
Because your interpreter is interacting with a delegate, not the underlying method.  The delegate signature should be the only signature that matters.  The fact that the delegate might point to a method that has params shouldn't matter.

**Reply by HaloFour** [reply]
Anyway, I guess it's up to the language team if they want to explicitly support that path.  Without explicitly including it into the language spec I would consider it as undefined.  You could make the argument that since private methods do include the attribute that it should apply to local functions as well, although the actual signature of a local function can differ from how it is declared based on what it does and how it's used.

**Reply by HaloFour** [reply]
Actually, as of C# 12, the natural type of the delegate would include the `params` parameters directly.  This is true for class methods, local functions and lambdas:

```cs
using System;
using System.Reflection;
using System.Linq;

public class Program {
    static void ClassMethod(params object[] Args) {

    }
    public static void Main() {
        static void LocalMethod(params object[] Args) {

        }

        var Lambda = (params object[] Args) => { };

        Console.WriteLine(((Delegate)ClassMethod).GetType().GetMethod("Invoke").GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
        Console.WriteLine(((Delegate)ClassMethod).GetType().GetMethod("Invoke").GetParameters()[0].GetCustomAttribute<ParamArrayAttribute>() is not null);
        Console.WriteLine(((Delegate)ClassMethod).GetType().GetMethod("Invoke").GetParameters()[0].CustomAttributes.Count());

        Console.WriteLine(((Delegate)LocalMethod).GetType().GetMethod("Invoke").GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
        Console.WriteLine(((Delegate)LocalMethod).GetType().GetMethod("Invoke").GetParameters()[0].GetCustomAttribute<ParamArrayAttribute>() is not null);
        Console.WriteLine(((Delegate)LocalMethod).GetType().GetMethod("Invoke").GetParameters()[0].CustomAttributes.Count());

        Console.WriteLine(Lambda.GetType().GetMethod("Invoke").GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
        Console.WriteLine(Lambda.GetType().GetMethod("Invoke").GetParameters()[0].GetCustomAttribute<ParamArrayAttribute>() is not null);
        Console.WriteLine(Lambda.GetType().GetMethod("Invoke").GetParameters()[0].CustomAttributes.Count());
    }
}
```

prints:

```
True
True
1
True
True
1
True
True
1
```

Although if you look at the target method for the lambda it also is not declared with `ParamArrayAttribute`.

**Reply by Joy-less** [reply]
That's very interesting, that it only works when reflecting through the method name of "Invoke". Thanks for the workaround!

**Comment by CyrusNajmabadi** [comment]
If you want to check if a delegate has a params array, you should be checking the `.Invoke` method of the *origina*l delegate type that you have declared.  The information about the actual method being passed along is not relevant.

**Reply by Joy-less** [reply]
What do you mean?

**Reply by CyrusNajmabadi** [reply]
I mean that the final pointer to the method being passed in isn't what should be examined. 

Instead, your code should be taking in some delegate type. Like:

```
delegate void JoyDel(params int[] ints);
void JoyMethod(JoyDel d) { ... }
...

JoyMethod(JoyLocalFunc);
```

Here you want to examine the JoyDel delegate type. Look at it's Invoke method to make the determination. Don't look at the JoyLocalFunc function. That's not relevant. It's an implementation detail.

**Reply by Joy-less** [reply]
That would work if the method signature was known at compile time, but it's not because it needs to support every method type. I already tried passing in `Func<T...>`s and `Action<T...>`s instead of `Delegate`s, but it's a very long process to support every overload of `Func<T...>`/`Action<T...>` and it also requires generics to be explicitly passed (which is annoying), for example:
```cs
public class Program {
	delegate void JoyDel<T1, T2, T3>(T1 a, T2 b, T3 c);
	public static void Main() {
		void MyExampleMethod(int a, bool b, string c) {
			
		}
		JoyMethod(MyExampleMethod); // The type arguments for method 'Program.JoyMethod<T1, T2, T3>(Program.JoyDel<T1, T2, T3>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
	}
	static void JoyMethod<T1, T2, T3>(JoyDel<T1, T2, T3> Thing) {
		
	}
}
```
Yes you could go `JoyMethod<int, bool, string>(MyExampleMethod);` but I don't want to do that, and it removes support for `Delegate`.
Also, this is still a bug, regardless of what I "should" be doing.

**Reply by HaloFour** [reply]
> Also, this is still a bug, regardless of what I "should" be doing.

That's up to the language team to decide.  If the spec doesn't state that the method should be emitted with those attributes, then it's not a bug.  Relying on internal implementation details via reflection is an inherently brittle strategy.

**Reply by CyrusNajmabadi** [reply]
> because it needs to support every method type

If you need to support every method type, then it's irrelevant how the original local function was declared. Just allow splatting of the final value was an array. 

> Also, this is still a bug, regardless of what I "should" be doing.

It is not. The original signature of a local function is a hidden detail. It can't be directly observed, and the language explicitly considers it a local. So the compiler is free to implement things as long as it is compliant with that.

**Comment by Joy-less** [comment]
I'd like to add that attributes are usually included for local function arguments (just not `ParamArrayAttribute`) making this a bug. Observe:

```cs
using System;
using System.Reflection;
using System.ComponentModel;
					
public class Program {
	public static void Main() {
		static void MethodA([DefaultValue("Default")] object[] Args) {
			
		}
		static void MethodB(params object[] Args) {
			
		}
		
		Console.WriteLine(((Delegate)MethodA).Method.GetParameters()[0].IsDefined(typeof(DefaultValueAttribute)));
		Console.WriteLine(((Delegate)MethodB).Method.GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
		Console.ReadLine();
	}
}
```
Output:
```
True
False
```

**Comment by Joy-less** [comment]
Thanks to @HaloFour for the workaround, but it only works with `ParamArrayAttribute`... bizarre!

```cs
using System;
using System.Reflection;
using System.ComponentModel;

public class Program {
    static void ClassMethod(params object[] Args) {

    }
    static void ClassMethod2([DefaultValue("Default")] object[] Args) {

    }
    public static void Main() {
	Console.WriteLine("Local ParamArrayAttribute");
        static void LocalMethod(params object[] Args) {

        }
        Console.WriteLine(((Delegate)LocalMethod).GetType().GetMethod("Invoke").GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
        Console.WriteLine(((Delegate)LocalMethod).Method.GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
		
	Console.WriteLine("\nLocal DefaultValueAttribute");
        static void LocalMethod2([DefaultValue("Default")] object[] Args) {

        }
        Console.WriteLine(((Delegate)LocalMethod2).GetType().GetMethod("Invoke").GetParameters()[0].IsDefined(typeof(DefaultValueAttribute)));
        Console.WriteLine(((Delegate)LocalMethod2).Method.GetParameters()[0].IsDefined(typeof(DefaultValueAttribute)));
		
	Console.WriteLine("\nClass ParamArrayAttribute");
        Console.WriteLine(((Delegate)ClassMethod).GetType().GetMethod("Invoke").GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
        Console.WriteLine(((Delegate)ClassMethod).Method.GetParameters()[0].IsDefined(typeof(ParamArrayAttribute)));
		
	Console.WriteLine("\nClass DefaultValueAttribute");
        Console.WriteLine(((Delegate)ClassMethod2).GetType().GetMethod("Invoke").GetParameters()[0].IsDefined(typeof(DefaultValueAttribute)));
        Console.WriteLine(((Delegate)ClassMethod2).Method.GetParameters()[0].IsDefined(typeof(DefaultValueAttribute)));
    }
}
```
Output (.NET 8):
```
Local ParamArrayAttribute
True
False

Local DefaultValueAttribute
False
True

Class ParamArrayAttribute
True
True

Class DefaultValueAttribute
False
True
```
